utils.py

```diff
@@ -235,4 +239,97 @@ def check_and_fill(args, target_len):
+    
+    Note:
+        This transform expects input data in the form of tuples (image, mask).
+        If only images are provided, it assumes no masks are present.
+    """
+    def __init__(self, aug, Num=2, **kwargs):
+        """
+        Initializes the SegmentationAlbumentationsTransform.
+
+        Args:
+            aug (callable): Albumentations augmentation function.
+            Num (float): Number of the augmented images minus from the batch size (default is 2).
+        """
+        super().__init__(**kwargs)
+        self.aug = aug
+        self.Num = Num
+
+    def encodes(self, x):
+        """
+        Applies Albumentations augmentations to input images and masks.
+
+        Args:
+            x (tuple or Tensor): Input data containing images and masks.
+
+        Returns:
+            Tensor or tuple of Tensors: Transformed images and masks (if provided).
+        """
+        try:
+            batch_img, batch_mask = x  # Expecting tuple (img, mask)
+        except ValueError:
+            batch_img = x  # Only one value provided, assuming it's just the image
+            batch_mask = None  # No mask provided
+            # Check if Num is greater than or equal to the batch size
+        if len(batch_img) <= self.Num:
+            raise ValueError(f"The Num parameter ({self.Num}) must be less than the batch size ({len(batch_img)}).")
+  
+        transformed_images = []
+        transformed_masks = []
+        
+        if batch_mask is None:
+            for img in batch_img:  # Ensure this iterates correctly over a batch
+                # Ensure the image has the correct dimensions [B, C, H, W] -> [B, H, W, C] for Albumentations
+                if img.dim() == 4:
+                    img_np = img.permute(0, 2, 3, 1).cpu().numpy()  # Change to [B, H, W, C]
+            
+                    # Apply augmentation to each image individually in the batch
+                    try:
+                        transformed = self.aug(image=img_np[0])  # Apply to the first (or only) image in the batch
+                        img_aug = np.transpose(transformed['image'], (2, 0, 1))
+                        transformed_images.append(TensorImage(torch.from_numpy(img_aug).unsqueeze(0)))  # Re-add batch dimension
+                    except Exception as e:
+                        print("Error during augmentation:", e)
+            return torch.stack(transformed_images)  # Stack to get [B, C, H, W]
+        
+        # Process each image and mask in the first proportion of the batch
+        for img, mask in zip(batch_img[:int(self.Num - len(batch_img))], batch_mask[:int(self.Num - len(batch_img))]):
+            # Normalize the image
+            img = img / img.max()
+        
+            # Permute the image dimensions from (C, H, W) to (H, W, C) for albumentations
+            img = img.permute(1, 2, 0)  # Now shape is [W, H, C]
+        
+            print('Applying augmentation')
```

Comment:
does this mean its printing for each batch that is transformed / not transformed (line 332)? If yes, I think this might be to verbose as there are a lot of print statements during training?

Reply:
ah right, I edited this line with the line in the train function .... 

       if transforms:
            print(f"Applying Augmentation on ({BATCH_SIZE - Num}) images from ({BATCH_SIZE}) images")
            # Use the imported aug_pipe
            transforms = SegmentationAlbumentationsTransform(aug_pipe, Num=Num)